File: Hash Table/HashTable.py
class HashTable:
    def __init__(self,size: int):
        self.nodes = [[] for i in range(size)]
        self.size = size
    
    def hash(self,key: str):
        hash_value = 0
        for c in key:
            hash_value = hash_value * 31 + ord(c)
        return hash_value % self.size

    def put(self,key: str, value: int):
        idx = self.hash(key)

        for i, (k,v) in enumerate(self.nodes[idx]):
            if k == key:
                self.nodes[idx][i] = (key,value)
                return
        self.nodes[idx].append((key,value))
    
    def get(self,key: str):
        idx = self.hash(key)

        for (k,v) in self.nodes[idx]:
            if k == key:
                return v
        
        print(f"Key {key} not found")
        return None

    def remove(self,key: str):
        idx = self.hash(key)

        for i,(k,v) in enumerate(self.nodes[idx]):
            if k == key:
                del self.nodes[idx][i]
    
    def print(self):
        for j, bucket in enumerate(self.nodes):
            print(f"Bucket {j+1}: [",end="")
            for k,v in bucket:
                print(f"({k}:{v}) -> ",end="")
            print("NULL ]")

Hash_Table = HashTable(5)
Hash_Table.put("William",21)
Hash_Table.put("Cesar",26)
Hash_Table.put("Silva",30)
Hash_Table.put("de",34)
Hash_Table.put("Carvalho",90)

Hash_Table.print()

File: Hash Table/HashTable.rs
use std::collections::LinkedList;

#[derive(Clone)] // Implementa o Clone para Entry
struct Entry {
    key: String,
    value: i32,
}

struct HashTable {
    buckets: Vec<LinkedList<Entry>>,
    size: usize,
}

impl HashTable {
    fn new(size_table: usize) -> Self {
        // Inicializa com listas vazias, não clonáveis
        Self {
            buckets: vec![LinkedList::new(); size_table],
            size: size_table,
        }
    }

    fn hash(&self, key: &str) -> usize {
        let mut hash_value: usize = 0;
        for c in key.chars() {
            hash_value = hash_value.wrapping_mul(31).wrapping_add(c as usize);
        }
        hash_value % self.size
    }

    fn put(&mut self, key: &str, value: i32) {
        let idx = self.hash(key);
        let bucket = &mut self.buckets[idx];

        for entry in bucket.iter_mut() {
            if entry.key == key {
                entry.value = value;
                return;
            }
        }

        bucket.push_back(Entry {
            key: key.to_string(),
            value,
        });
    }

    fn get(&self, key: &str) -> Option<i32> {
        let idx = self.hash(key);
        let bucket = &self.buckets[idx];

        for entry in bucket {
            if entry.key == key {
                return Some(entry.value);
            }
        }
        None
    }

    fn remove(&mut self, key: &str) {
        let idx = self.hash(key);
        let bucket = &mut self.buckets[idx];

        // Remoção manual do item
        let mut prev: Option<&mut Entry> = None;
        let mut to_remove = None;

        for entry in bucket.iter_mut() {
            if entry.key == key {
                to_remove = Some(entry);
                break;
            }
            prev = Some(entry);
        }

        if let Some(entry) = to_remove {
            if let Some(prev_entry) = prev {
                // Remover o item da lista
                *prev_entry = entry.clone(); // Agora Entry é clonável
            } else {
                // Se for o primeiro item, remove diretamente
                bucket.pop_front();
            }
        }
    }

    fn print(&self) {
        for (i, bucket) in self.buckets.iter().enumerate() {
            print!("Bucket {}: ", i + 1);
            for entry in bucket {
                print!("({}: {}) -> ", entry.key, entry.value);
            }
            println!("NULL");
        }
    }
}

fn main() {
    let mut table = HashTable::new(10);

    table.put("William", 21);
    table.put("Cesar", 26);
    table.put("Silva", 30);
    table.put("De", 34);
    table.put("Carvalho", 90);

    table.print();

    println!("\nBuscando valores:");
    println!("William: {:?}", table.get("William"));
    println!("Cesar: {:?}", table.get("Cesar"));
    println!("Silva: {:?}", table.get("Silva"));

    println!("\nRemovendo 'De'...");
    table.remove("De");

    table.print();
}
